[fake_pg] allow fake_pg allgather to do some simple validation

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #104213

Note that in general it's not good form to try to make FakePG work with 'real data',
but the reasoning here is that we want FakePG to work with DeviceMesh's init code
that have the data validation, which makes it worth the tradeoff.

In general user should use MTPG or normal PG for cases where they may care about
real data from collectives

wanchaol committed:
[fake_pg] allow fake_pg allgather to do some simple validation

[ghstack-poisoned]

wconstab commented on torch/testing/_internal/distributed/fake_pg.py:
```diff
@@ -44,6 +44,13 @@ def reduce_scatter(self, output_tensor, scatter_list, opts=ReduceScatterOptions(
+        # assume each rank have the same input tensor so we just copy to the results
+        # since it's not a real allgather, we simply make this copying logic to let
+        # some simple validation works (i.e. calling allgather to see if each rank have
+        # the same tensor or not)
```
can you make a more explicit comment here and also in the PR desc / commit msg, explaining that in general its not good form to try to make FakePG work with 'real data', but the reasoning here is that we want FakePG to work with DeviceMesh's initialization code which we feel is worth this tradeoff?  

(And suggest in general people should use MTPG for cases where they may care about real data from collectives)

wanchaol committed:
Update on "[fake_pg] allow fake_pg allgather to do some simple validation"


Note that in general it's not good form to try to make FakePG work with 'real data',
but the reasoning here is that we want FakePG to work with DeviceMesh's init code
that have the data validation, which makes it worth the tradeoff.

In general user should use MTPG or normal PG for cases where they may care about
real data from collectives

[ghstack-poisoned]

wanchaol commented:
@pytorchbot merge
